File: engine/events.py
"""
Event types for the game engine.

Events form a typed log of everything that happens during the game.
They are designed to be consumed by UI layers for rendering animations
and game state updates.
"""

from dataclasses import dataclass

from engine.types import (
    PlayerId,
    LocationIndex,
    TurnNumber,
    InstanceId,
    Power,
    Energy,
    GameResult,
)


# =============================================================================
# Turn Events
# =============================================================================


@dataclass(frozen=True, slots=True)
class TurnStartedEvent:
    """A new turn has begun."""

    turn: TurnNumber
    """The turn number that started."""

    @property
    def event_type(self) -> str:
        return "turn_started"


@dataclass(frozen=True, slots=True)
class TurnEndedEvent:
    """A turn has ended."""

    turn: TurnNumber
    """The turn number that ended."""

    @property
    def event_type(self) -> str:
        return "turn_ended"


# =============================================================================
# Energy Events
# =============================================================================


@dataclass(frozen=True, slots=True)
class EnergySetEvent:
    """Player energy was set."""

    player_id: PlayerId
    """The player whose energy changed."""

    energy: Energy
    """The new energy value."""

    @property
    def event_type(self) -> str:
        return "energy_set"


@dataclass(frozen=True, slots=True)
class EnergySpentEvent:
    """Player spent energy."""

    player_id: PlayerId
    """The player who spent energy."""

    amount: Energy
    """The amount spent."""

    remaining: Energy
    """Remaining energy after spending."""

    @property
    def event_type(self) -> str:
        return "energy_spent"


# =============================================================================
# Card Events
# =============================================================================


@dataclass(frozen=True, slots=True)
class CardDrawnEvent:
    """A card was drawn from deck to hand."""

    player_id: PlayerId
    """The player who drew the card."""

    card_instance_id: InstanceId
    """The instance ID of the drawn card."""

    @property
    def event_type(self) -> str:
        return "card_drawn"


@dataclass(frozen=True, slots=True)
class CardPlayedEvent:
    """A card was played to a location."""

    player_id: PlayerId
    """The player who played the card."""

    card_instance_id: InstanceId
    """The instance ID of the played card."""

    location: LocationIndex
    """The location where the card was played."""

    @property
    def event_type(self) -> str:
        return "card_played"


@dataclass(frozen=True, slots=True)
class CardRevealedEvent:
    """A card was revealed at a location."""

    card_instance_id: InstanceId
    """The instance ID of the revealed card."""

    location: LocationIndex
    """The location where the card was revealed."""

    player_id: PlayerId
    """The player who owns the card."""

    @property
    def event_type(self) -> str:
        return "card_revealed"


@dataclass(frozen=True, slots=True)
class CardMovedEvent:
    """A card was moved between locations."""

    card_instance_id: InstanceId
    """The instance ID of the moved card."""

    from_location: LocationIndex
    """The location the card moved from."""

    to_location: LocationIndex
    """The location the card moved to."""

    source_card_id: InstanceId | None
    """The card that caused this move (if any)."""

    @property
    def event_type(self) -> str:
        return "card_moved"


@dataclass(frozen=True, slots=True)
class CardDestroyedEvent:
    """A card was destroyed."""

    card_instance_id: InstanceId
    """The instance ID of the destroyed card."""

    location: LocationIndex
    """The location where the card was destroyed."""

    source_card_id: InstanceId | None
    """The card that caused this destruction (if any)."""

    @property
    def event_type(self) -> str:
        return "card_destroyed"


# =============================================================================
# Power Events
# =============================================================================


@dataclass(frozen=True, slots=True)
class PowerChangedEvent:
    """A card's power was modified."""

    card_instance_id: InstanceId
    """The instance ID of the affected card."""

    old_power: Power
    """The power before the change."""

    new_power: Power
    """The power after the change."""

    source_card_id: InstanceId | None
    """The card that caused this change (if any)."""

    @property
    def event_type(self) -> str:
        return "power_changed"


# =============================================================================
# Action Events
# =============================================================================


@dataclass(frozen=True, slots=True)
class PlayerPassedEvent:
    """A player passed their turn."""

    player_id: PlayerId
    """The player who passed."""

    @property
    def event_type(self) -> str:
        return "player_passed"


@dataclass(frozen=True, slots=True)
class ActionInvalidEvent:
    """An action was invalid and rejected."""

    player_id: PlayerId
    """The player whose action was invalid."""

    reason: str
    """Explanation of why the action was invalid."""

    @property
    def event_type(self) -> str:
        return "action_invalid"


# =============================================================================
# Game Events
# =============================================================================


@dataclass(frozen=True, slots=True)
class GameStartedEvent:
    """The game has started."""

    @property
    def event_type(self) -> str:
        return "game_started"


@dataclass(frozen=True, slots=True)
class GameEndedEvent:
    """The game has ended."""

    result: GameResult
    """The final result of the game."""

    location_winners: tuple[PlayerId | None, PlayerId | None, PlayerId | None]
    """Winner of each location (None = tie)."""

    location_powers: tuple[
        tuple[Power, Power],
        tuple[Power, Power],
        tuple[Power, Power],
    ]
    """Power totals at each location: ((p0_loc0, p1_loc0), ...)."""

    total_power: tuple[Power, Power]
    """Total power across all locations for each player."""

    @property
    def event_type(self) -> str:
        return "game_ended"


# =============================================================================
# Event Union Type
# =============================================================================

GameEvent = (
    TurnStartedEvent
    | TurnEndedEvent
    | EnergySetEvent
    | EnergySpentEvent
    | CardDrawnEvent
    | CardPlayedEvent
    | CardRevealedEvent
    | CardMovedEvent
    | CardDestroyedEvent
    | PowerChangedEvent
    | PlayerPassedEvent
    | ActionInvalidEvent
    | GameStartedEvent
    | GameEndedEvent
)

File: web/src/engine/events.ts
/**
 * Game events for the event log.
 * 
 * Ported from Python engine/events.py
 * These events drive animations in the UI.
 */

import type {
  PlayerId,
  LocationIndex,
  InstanceId,
  TurnNumber,
  Energy,
  Power,
  GameResult,
} from './types';

// =============================================================================
// Event Types (Discriminated Union)
// =============================================================================

export interface GameStartedEvent {
  readonly type: 'GameStarted';
}

export interface GameEndedEvent {
  readonly type: 'GameEnded';
  readonly result: GameResult;
  readonly locationWinners: readonly (PlayerId | null)[];
  readonly locationPowers: readonly (readonly [Power, Power])[];
  readonly totalPower: readonly [Power, Power];
}

export interface TurnStartedEvent {
  readonly type: 'TurnStarted';
  readonly turn: TurnNumber;
}

export interface TurnEndedEvent {
  readonly type: 'TurnEnded';
  readonly turn: TurnNumber;
}

export interface EnergySetEvent {
  readonly type: 'EnergySet';
  readonly playerId: PlayerId;
  readonly energy: Energy;
}

export interface BonusEnergyEvent {
  readonly type: 'BonusEnergy';
  readonly playerId: PlayerId;
  readonly bonus: Energy;
  readonly locationsWon: number;
  readonly newTotal: Energy;
}

export interface EnergySpentEvent {
  readonly type: 'EnergySpent';
  readonly playerId: PlayerId;
  readonly amount: Energy;
  readonly remaining: Energy;
}

export interface CardDrawnEvent {
  readonly type: 'CardDrawn';
  readonly playerId: PlayerId;
  readonly cardInstanceId: InstanceId;
}

export interface CardPlayedEvent {
  readonly type: 'CardPlayed';
  readonly playerId: PlayerId;
  readonly cardInstanceId: InstanceId;
  readonly location: LocationIndex;
}

export interface CardRevealedEvent {
  readonly type: 'CardRevealed';
  readonly cardInstanceId: InstanceId;
  readonly location: LocationIndex;
  readonly playerId: PlayerId;
}

export interface CardMovedEvent {
  readonly type: 'CardMoved';
  readonly cardInstanceId: InstanceId;
  readonly fromLocation: LocationIndex;
  readonly toLocation: LocationIndex;
  readonly sourceCardId: InstanceId;
}

export interface CardDestroyedEvent {
  readonly type: 'CardDestroyed';
  readonly cardInstanceId: InstanceId;
  readonly location: LocationIndex;
  readonly sourceCardId: InstanceId;
}

export interface PowerChangedEvent {
  readonly type: 'PowerChanged';
  readonly cardInstanceId: InstanceId;
  readonly oldPower: Power;
  readonly newPower: Power;
  readonly sourceCardId: InstanceId;
}

export interface PlayerPassedEvent {
  readonly type: 'PlayerPassed';
  readonly playerId: PlayerId;
}

export interface ActionInvalidEvent {
  readonly type: 'ActionInvalid';
  readonly playerId: PlayerId;
  readonly reason: string;
}

// Union of all event types
export type GameEvent =
  | GameStartedEvent
  | GameEndedEvent
  | TurnStartedEvent
  | TurnEndedEvent
  | EnergySetEvent
  | BonusEnergyEvent
  | EnergySpentEvent
  | CardDrawnEvent
  | CardPlayedEvent
  | CardRevealedEvent
  | CardMovedEvent
  | CardDestroyedEvent
  | PowerChangedEvent
  | PlayerPassedEvent
  | ActionInvalidEvent;

// =============================================================================
// Animation Configuration per Event Type
// =============================================================================

export interface AnimationConfig {
  readonly duration: number; // seconds
  readonly delay?: number;
  readonly interruptible: boolean;
}

export const EVENT_ANIMATIONS: Record<GameEvent['type'], AnimationConfig> = {
  GameStarted: { duration: 0.5, interruptible: true },
  GameEnded: { duration: 1.0, interruptible: false },
  TurnStarted: { duration: 0.3, interruptible: true },
  TurnEnded: { duration: 0.2, interruptible: true },
  EnergySet: { duration: 0.3, interruptible: true },
  BonusEnergy: { duration: 0.4, interruptible: true },
  EnergySpent: { duration: 0.2, interruptible: true },
  CardDrawn: { duration: 0.4, interruptible: true },
  CardPlayed: { duration: 0.5, interruptible: false },
  CardRevealed: { duration: 0.6, interruptible: false },
  CardMoved: { duration: 0.7, interruptible: false },
  CardDestroyed: { duration: 0.5, interruptible: false },
  PowerChanged: { duration: 0.4, interruptible: true },
  PlayerPassed: { duration: 0.2, interruptible: true },
  ActionInvalid: { duration: 0.3, interruptible: true },
};
